post-mortems/2019-09-03-discovery-elasticsearch-deletion.md

```diff
@@ -0,0 +1,54 @@
+
+Another possible cause may be an hardware or software issue. The domain uses two nodes. Under Instance Health, there are six Data Instances:
+   * Elasticsearch-EjrqmHyDS-qK0MJvH9A2Mg: went quiet on Aug 31 at/after 9:00am
+   * Elasticsearch-u3AewBjlSD2rwrPdRbI_yg: went quiet on Aug 31 at/after 8:00am
+   * Elasticsearch-xWtfBlWnQiG3kYR1b5qYUw: live from Aug 31 10am to 9.03 8pm
+   * Elasticsearch--67y18NIR8SGpGfWFzgcPQ: [ditto above]
+   * Elasticsearch-zigD5d_8SbmTKCvN9bPozA: live from 9.03 8pm to current
+   * Elasticsearch-7ticgp4RRgiJBirouz8fFA: [ditto above]
+
+I don't know what causes data nodes to be retired and replaced. It's suspect that two new nodes were activated about an hour before the outage. It seems possible that a routine process to swap in new nodes simply failed to initialize them with the right data. It's also curious that two new nodes were activated at 8pm on 9.03. I believe this coincides with [enabling](https://console.aws.amazon.com/es/home?region=us-east-1#domain:resource=discovery-api-production;action=dashboard) the [error logs](https://console.aws.amazon.com/cloudwatch/home?region=us-east-1#logEventViewer:group=/aws/aes/domains/discovery-api-production/application-logs) for the domain. Shortly after that was enabled, I see log entries containing "failed to connect to node {-67y18N}" and "failed to connect to node {xWtfBlW}{xWtfBlWnQiG3kYR1b5qYUw..." suggesting those nodes' health degraded, causing them to be swapped. The reason for the degradation could have been the sheer number of failed queries hitting the index (due to absence of a field mapping) or could be a side effect of enabling logging. In any case, the nodes created on 9.03 at 8pm to replace the failed nodes have remained without being swapped out (including during the brief period on Sep 3 from about 8pm to 10:41pm when the DiscoveryApi EB continued to send failing queries to the old index, which troubles the theory that it was the queries themselves that destabilized the nodes).
```

Comment:
> I believe this coincides with [enabling](https://console.aws.amazon.com/es/home?region=us-east-1#domain:resource=discovery-api-production;action=dashboard) the [error logs](https://console.aws.amazon.com/cloudwatch/home?region=us-east-1#logEventViewer:group=/aws/aes/domains/discovery-api-production/application-logs) for the domain.

That's part of the replacement of nodes? I.e., that's not a person going in and enabling error logs, right?

Reply:
I mean to suggest the replacement of nodes may occur *because* of the configuration change (i.e. enabling logs), but I I'm not sure. The first messages in the error logs were "failed to connect to node.." messages. I'm not sure if those error messages are naturally going to appear any time you change a major configuration in the domain (like enabling error logging) or if they reflected some other instability. Total node replacement occurs at odd times over the lifetime of the domain. In this case it appeared to occur very close to the time I made a config change.